#498: Add support for JSON.OBJKEYS

iamskp11 opened the pull request:
### **Issue**
Issue fixed : https://github.com/DiceDB/dice/issues/498

### **PR Description**
This PR implements JSON.OBJKEYS command, and add unit tests, integration tests and benchmarking evaluation for the command.

### **Test Results**
1. Dice DB vs Redis
![127 0 0 17379](https://github.com/user-attachments/assets/74418ea5-2535-4a5c-8669-af6c80ebc31c)
Dice DB CLI usage

![127 0 0 16379 json objkeys doc](https://github.com/user-attachments/assets/41b0d28d-a48f-4d0c-9ddd-eec6d5ef738e)
Redis CLI usage

2. Unit Tests
<img width="883" alt="Pasted Graphic 3" src="https://github.com/user-attachments/assets/05a336c5-f649-4920-b245-2a5bfb89946d">
<img width="741" alt="Pasted Graphic 4" src="https://github.com/user-attachments/assets/a2a58ba6-46c9-4ec1-beb9-74bde859198c">

3. Integration Tests
<img width="841" alt="Pasted Graphic 2" src="https://github.com/user-attachments/assets/6a0b9ef5-90ab-4352-b450-4c6792b86e52">

4. Benchmarking
<img width="1017" alt="Pasted Graphic 5" src="https://github.com/user-attachments/assets/ee84af82-a531-4199-a9da-46af6cd28b15">

iamskp11 commented:
Hey @lucifercr07 @JyotinderSingh @jujiale , requesting for your reviews on this PR.

jujiale commented on internal/eval/eval.go:
```diff
@@ -3553,6 +3553,82 @@ func evalJSONNUMINCRBY(args []string, store *dstore.Store) []byte {
+		path = args[1]
+	}
+
+	// Retrieve the object from the database
+	// TODO: Check if we should throw error here
+	obj := store.Get(key)
+	if obj == nil {
+		return clientio.RespNIL
+	}
```
thanks for your contribution. in redis. if the key is not exist. should return an error.
`

      127.0.0.1:6379> JSON.SET doc $ '{"a":[3], "nested": {"a": {"b":2, "c": 1}}}'
      OK
      127.0.0.1:6379> JSON.OBJKEYS docs $..a
      (error) ERR could not perform this operation on a key that doesn't exist

`

iamskp11 replied:
In this [commit](https://github.com/DiceDB/dice/pull/629/commits/946439cf099bc825c1da2bc6b445b2cc1af2328d), 
1. Error will be thrown for non existent key

iamskp11 replied:
At the next step post getting value of the key, we check if it is json and throw appropriate error So, its handled already.
<img width="543" alt="image" src="https://github.com/user-attachments/assets/3fddeb6f-ae16-436d-a896-2c46b645b718">

jujiale commented on internal/eval/eval.go:
```diff
@@ -3553,6 +3553,82 @@ func evalJSONNUMINCRBY(args []string, store *dstore.Store) []byte {
+
+	// Execute the JSONPath query
+	results := expr.Get(jsonData)
+	if len(results) == 0 {
```
if the jsonpath query is empty. should return an empty array.
`

    127.0.0.1:6379> JSON.SET doc $ '{"a":[3], "nested": {"a": {"b":2, "c": 1}}}'
    OK
    127.0.0.1:6379> JSON.OBJKEYS doc $..ddd
    (empty array)


`

iamskp11 replied:
In this [commit](https://github.com/DiceDB/dice/pull/629/commits/946439cf099bc825c1da2bc6b445b2cc1af2328d), 
1. Changed the response to return Empty Array if jsonpath query is empty.

jujiale commented:
thanks for your contribution. this command logic looks well. I left a few comments. please have a look. thanks

iamskp11 commented:
> thanks for your contribution. this command logic looks well. I left a few comments. please have a look. thanks

Addressed all comments in this [commit](https://github.com/DiceDB/dice/pull/629/commits/946439cf099bc825c1da2bc6b445b2cc1af2328d)

iamskp11 commented:
Hey @JyotinderSingh @lucifercr07 , requesting you to have a look and review this PR. Have addressed initial reviews.

JyotinderSingh reviewed (approved):
Thanks for adding support for JSON.OBJKEYS @iamskp11. Thanks for the helpful reviews @jujiale!

The changes look good, approved.

manishrw reviewed (commented):
@iamskp11 please take these changes up. Let's maintain high bar of quality

iamskp11 commented on internal/eval/eval.go:
```diff
@@ -3621,6 +3621,81 @@ func evalJSONNUMINCRBY(args []string, store *dstore.Store) []byte {
+	// Retrieve the object from the database
+	obj := store.Get(key)
+	if obj == nil {
+		return diceerrors.NewErrWithMessage("could not perform this operation on a key that doesn't exist")
```
Thought was to make JSON.OBJKEYS consistent with Redis. Redis throws this error when we try to query with an unavailable key.
<img width="528" alt="image" src="https://github.com/user-attachments/assets/f7a907b7-3334-4261-8e91-e0fd1f1070b1"> (Redis)

Also, if we don't pass rootpath , i.e. '$', we get (nil) as response in Redis. In DiceDB, I observed in all commands, we consider rootPath ($) as defaultPath. So, even if we don't pass any path, we respond with same key DNE error. 

<img width="575" alt="image" src="https://github.com/user-attachments/assets/241cf459-f50b-47a0-8af9-9f5219a09a7d"> (DiceDB)


I am not sure why absence and presence of path is leading to different response in case of missing keys in case of Redis. 
<img width="546" alt="image" src="https://github.com/user-attachments/assets/6d439737-fe3e-4c75-a36f-be44c7e5b15e"> (Redis)

Please suggest if we should be responding (nil) in case of no path, and unavailable key.

iamskp11 commented:
> @iamskp11 please take these changes up. Let's maintain high bar of quality

Hey @manishrw Thanks for reviewing! Have responded to one of the comment, please have a look and tell your views. Next, I will be creating a new PR with your suggested changes.

iamskp11 commented on internal/eval/eval.go:
```diff
@@ -3621,6 +3621,81 @@ func evalJSONNUMINCRBY(args []string, store *dstore.Store) []byte {
+	// Check if the object is of JSON type
+	errWithMessage := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeJSON, object.ObjEncodingJSON)
+	if errWithMessage != nil {
+		return errWithMessage
```
AssertTypeAndEncoding function already returns formatted Error Message. So, we can directly return that. I think converting the response(which is byte array) again to string will be redundant.
<img width="752" alt="image" src="https://github.com/user-attachments/assets/f03d5c85-2c3f-4180-bab3-60adbadf1eae">
